src/firestore.rules.test.js

```diff
@@ -157,10 +163,290 @@ test('allows scholarship delete when user is admin', async () => {
+
+  test('fails update when author id not a string', async () => {
+    await assertSucceeds(collection('ai').set(newScholarship));
+    return assertFails(
+      collection('ai').set({
+        ...newScholarship,
+        author: { id: 2, email: 'test@gmail.com' },
+      })
+    );
+  });
+
+  test('fails create when author email not a string', () => {
+    return assertFails(
+      collection('ae').set({
+        ...newScholarship,
+        author: { email: 2, id: aliceId },
+      })
+    );
+  });
+
+  test('fails update when author email not a string', async () => {
+    await assertSucceeds(collection('ae').set(newScholarship));
+    return assertFails(
+      collection('ae').set({
+        ...newScholarship,
+        author: { id: aliceId, email: 2 },
+      })
+    );
+  });
+
+  describe('requirements validation', () => {
+    const scholarship = { ...newScholarship };
+    scholarship.requirements = {};
+
+    test('fails create when requirements not an object', () => {
+      return assertFails(
+        collection('re').set({ ...newScholarship, requirements: 3.4 })
+      );
+    });
+
+    test('fails update when requirements not an object', async () => {
+      await assertSucceeds(collection('re').set(newScholarship));
+      return assertFails(
+        collection('re').set({ ...newScholarship, requirements: 3.4 })
+      );
+    });
+
+    test('fails create when gpa is not a number', () => {
+      return assertFails(
+        collection('gp').set({
+          ...newScholarship,
+          requirements: { gpa: '3.0' },
+        })
+      );
+    });
+
+    test('fails update when gpa is not a number', async () => {
+      await assertSucceeds(collection('gp').set(newScholarship));
+      return assertFails(
+        collection('gp').set({
+          ...newScholarship,
+          requirements: { gpa: '3.0' },
+        })
+      );
+    });
+
+    test('fails create when gpa is less than 0', () => {
+      return assertFails(
+        collection('gp').set({ ...newScholarship, requirements: { gpa: -3 } })
+      );
+    });
+
+    test('fails update when gpa is less than 0', async () => {
+      await assertSucceeds(collection('gp').set(newScholarship));
+      return assertFails(
+        collection('gp').set({
+          ...newScholarship,
+          requirements: { gpa: -3 },
+        })
+      );
+    });
+
+    test('fails create when gpa is greater than 4', () => {
+      return assertFails(
+        collection('gp').set({ ...newScholarship, requirements: { gpa: 4.1 } })
+      );
+    });
+
+    test('fails update when gpa is greater than 4', async () => {
+      await assertSucceeds(collection('gp').set(newScholarship));
+      return assertFails(
+        collection('gp').set({ ...newScholarship, requirements: { gpa: 4.1 } })
+      );
+    });
+
+    test('fails create when ethnicities not an array', () => {
+      scholarship.requirements.ethnicities = '[]';
+      return assertFails(collection.set(scholarship));
```

Comment:
`scholarship` seems to be undefined, do you mean to do `{...newScholarship, requirements: {ethnicities: '[]'}`?

Reply:
I got rid of the global variable and ended up doing this